src/display/pattern_helper.js

```diff
@@ -37,21 +37,57 @@ ShadingIRs.RadialAxial = {
     const p1 = raw[5];
     const r0 = raw[6];
     const r1 = raw[7];
+    const matrix = raw[8];
+
     return {
-      getPattern: function RadialAxial_getPattern(ctx) {
-        applyBoundingBox(ctx, bbox);
+      getPattern: function RadialAxial_getPattern(ctx, owner, shadingFill) {
+        const tmpCanvas = owner.cachedCanvases.getCanvas(
+          "pattern",
+          ctx.canvas.width,
+          ctx.canvas.height,
+          true
+        );
+
+        const tmpCtx = tmpCanvas.context;
+        tmpCtx.clearRect(0, 0, tmpCtx.canvas.width, tmpCtx.canvas.height);
+        tmpCtx.beginPath();
+        tmpCtx.rect(0, 0, tmpCtx.canvas.width, tmpCtx.canvas.height);
+
+        if (!shadingFill) {
+          tmpCtx.setTransform.apply(tmpCtx, owner.baseTransform);
+          if (matrix) {
+            tmpCtx.transform.apply(tmpCtx, matrix);
+          }
+        } else {
+          tmpCtx.setTransform.apply(tmpCtx, ctx.mozCurrentTransform);
+        }
+        applyBoundingBox(tmpCtx, bbox);
+
         let grad;
         if (type === "axial") {
-          grad = ctx.createLinearGradient(p0[0], p0[1], p1[0], p1[1]);
+          grad = tmpCtx.createLinearGradient(p0[0], p0[1], p1[0], p1[1]);
         } else if (type === "radial") {
-          grad = ctx.createRadialGradient(p0[0], p0[1], r0, p1[0], p1[1], r1);
+          grad = tmpCtx.createRadialGradient(
+            p0[0],
+            p0[1],
+            r0,
+            p1[0],
+            p1[1],
+            r1
+          );
         }
 
         for (let i = 0, ii = colorStops.length; i < ii; ++i) {
           const c = colorStops[i];
           grad.addColorStop(c[0], c[1]);
         }
-        return grad;
+
+        tmpCtx.fillStyle = grad;
+        tmpCtx.fill();
+
+        const pattern = ctx.createPattern(tmpCanvas.canvas, "repeat");
+        pattern.setTransform(new DOMMatrix(ctx.mozCurrentTransformInverse));
```

Comment:
I *really* don't want to complicate this patch in any way (especially for the `MOZCENTRAL` build), given how much it improves the pattern handling, but are we worried about people using the general PDF.js library in older Firefox versions?

---

According to https://developer.mozilla.org/en-US/docs/Web/API/CanvasPattern/setTransform#browser_compatibility, using a `DOMMatrix` as the argument has only been supported since Firefox 79 (the current ESR is still on the 78 branch).